src/software/embedded/robot_diagnostics_cli/embedded_communication.py

```diff
@@ -0,0 +1,153 @@
+
+        self.send_estop_state_thread = threading.Thread(
+            target=self.__update_estop_state, daemon=True
+        )
+
+        self.previous_estop_state = False
+
+        # only checks for estop if we are in physical estop mode
+        if self.estop_mode == EstopMode.PHYSICAL_ESTOP:
+            try:
+                self.estop_reader = tbots_cpp.ThreadedEstopReader(
+                    self.estop_path, 115200
+                )
+            except Exception as e:
+                raise Exception(f"Invalid Estop found at location {self.estop_path} as {e}")
+
+    def __receive_robot_status(self, robot_status: Message) -> None:
+        """Forwards the given robot status to the full system along with the round-trip time
+        :param robot_status: RobotStatus to extract information from
+        """
+        self.epoch_timestamp_seconds = robot_status.time_sent.epoch_timestamp_seconds
+        self.battery_voltage = robot_status.power_status.battery_voltage
+        self.primitive_packet_loss_percentage = robot_status.network_status.primitive_packet_loss_percentage
+        self.primitive_executor_step_time_ms = robot_status.thunderloop_status.primitive_executor_step_time_ms
+
+    def __should_send_packet(self) -> bool:
+        """Returns True if should send a proto
+        :return: boolean
+        """
+        return ((self.estop_mode != EstopMode.DISABLE_ESTOP) and self.estop_is_playing)
+
+    def __send_primitive_set(self, primitive: Primitive) -> None:
+        """Forward PrimitiveSet protos from diagnostics to the robots."""
+        self.robot_primitives[self.robot_id] = primitive
+        # initialize total primitive set and send it
+        primitive_set = PrimitiveSet(
+            time_sent=Timestamp(epoch_timestamp_seconds=time.time()),
+            stay_away_from_ball=False,
+            robot_primitives=(
+                self.robot_primitives
+                if not self.should_send_stop
+                else {
+                    robot_id: Primitive(stop=StopPrimitive())
+                    for robot_id in self.robot_primitives.keys()
+                }
+            ),
+            sequence_number=self.sequence_number,
+        )
+
+        self.sequence_number += 1
+
+        if self.__should_send_packet() or self.should_send_stop:
+            self.send_primitive_set.send_proto(primitive_set)
+            self.should_send_stop = False
+
+    def __update_estop_state(self) -> None:
+        """Updates the current estop status proto if estop is not disabled
+        Always in physical estop mode, uses the physical estop value
+        """
+        while True:
+            if self.estop_mode == EstopMode.DISABLE_ESTOP:
+                self.should_send_stop = False
+            elif self.estop_mode == EstopMode.PHYSICAL_ESTOP:
+                self.should_send_stop = not self.estop_reader.isEstopPlay()
+            else:
+                self.should_send_stop = True
```

Comment:
needs a sleep, otherwise this thread will unnecessarily slow down your code:

```suggestion
        while True:
            self.should_send_stop = not self.estop_reader.isEstopPlay() if self.estop_mode = EstopMode.PHYSICAL_ESTOP else False
            time.sleep(0.1)
```